File: app/src/main/java/jp/ac/u_tokyo/t/seo/station/app/AnimationView.java
package jp.ac.u_tokyo.t.seo.station.app;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.Rect;
import android.os.Parcel;
import android.os.Parcelable;
import android.os.SystemClock;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewTreeObserver;

/**
 * @author Seo-4d696b75
 * @version 2019/04/18.
 */

public class AnimationView extends View{

    public AnimationView(Context context){
        this(context, null);
    }

    public AnimationView(Context context, AttributeSet set){
        this(context, set, 0);
    }

    public AnimationView(Context context, AttributeSet set, int defaultAttr){
        super(context, set, defaultAttr);
        getViewTreeObserver().addOnGlobalLayoutListener(LAYOUT_LISTENER);
        mPaint = new Paint();
        mMatrix = new Matrix();
        Resources res = context.getResources();
        mForeground = BitmapFactory.decodeResource(res, R.drawable.launch_icon_old);
        mBackground = BitmapFactory.decodeResource(res, R.drawable.loop_line);
        setWillNotDraw(false);
    }

    private final ViewTreeObserver.OnGlobalLayoutListener LAYOUT_LISTENER = new ViewTreeObserver.OnGlobalLayoutListener(){
        @Override
        public void onGlobalLayout(){
            getViewTreeObserver().removeOnGlobalLayoutListener(LAYOUT_LISTENER);
            mWidth = getWidth();
            mHeight = getHeight();
            mSrc = new Rect(0,0,mForeground.getWidth(),mForeground.getHeight());
            mDes = new Rect(
                    (int)(mWidth * 0.1),
                    (int)(mHeight * 0.13),
                    (int)(mWidth * 0.8),
                    (int)(mHeight * 0.83)
            );
            runAnimation(mRequestRunning);
        }
    };

    private int mWidth, mHeight;
    private long mTime;
    private float mDegree;
    private boolean mRunning;
    private boolean mRequestRunning;
    private boolean mRequestStop;
    private Bitmap mBackground, mForeground;
    private Paint mPaint;
    private Matrix mMatrix;
    private Rect mSrc, mDes;

    public void runAnimation(boolean run){
        if ( mWidth == 0 || mHeight == 0 || mBackground == null || mForeground == null ){
            mRequestRunning = run;
            return;
        }
        if ( mRunning != run ){
            if ( mRunning ){
                mRequestStop = true;
            }else{
                mRunning = true;
                invalidate();
            }
        }
    }

    @Override
    protected void onDraw(Canvas canvas){
        if ( mWidth == 0 || mHeight == 0 || mBackground == null || mForeground == null ) return;
        long time = SystemClock.uptimeMillis();
        if ( mRunning && mTime != 0 ){
            float degree = mDegree + (time - mTime) * 0.01f;
            if ( mRequestStop ){
                float step = 360f / 16;
                float threshold = (float)Math.ceil(mDegree/step) * step;
                if ( degree >= threshold ){
                    mRunning = false;
                    mRequestStop = false;
                    mDegree = 0;
                    time = 0;
                }
            }
            if ( degree > 360 ) degree -= 360;
            mDegree = degree;
        }
        mMatrix.reset();
        mMatrix.postRotate(mDegree, mBackground.getWidth()/2f, mBackground.getHeight()/2f);
        mMatrix.postScale((float)mWidth/mBackground.getWidth(), (float)mHeight/mBackground.getHeight());
        //canvas.drawColor(0, PorterDuff.Mode.CLEAR);
        canvas.drawBitmap(mBackground, mMatrix, mPaint);
        canvas.drawBitmap(mForeground, mSrc, mDes, mPaint);

        mTime = time;
        if ( mRunning ) invalidate();
    }

    @Override
    protected void onRestoreInstanceState(Parcelable source){
        SavedState state = (SavedState)source;
        super.onRestoreInstanceState(state.getSuperState());

        runAnimation(state.mRunning);


    }

    @Override
    protected Parcelable onSaveInstanceState(){
        Parcelable superState = super.onSaveInstanceState();
        SavedState state = new SavedState(superState);
        state.mRunning = this.mRunning;
        return state;
    }

    private static class SavedState extends BaseSavedState {

        boolean mRunning;

        SavedState(Parcelable state){
            super(state);
        }


        SavedState(Parcel source){
            super(source);
            boolean[] array = new boolean[1];
            source.readBooleanArray(array);
            mRunning = array[0];
        }

        @Override
        public void writeToParcel(final Parcel out, final int flags) {
            super.writeToParcel(out, flags);
            boolean[] array = new boolean[]{mRunning};
            out.writeBooleanArray(array);
        }

        public static final Parcelable.Creator<SavedState> CREATOR = new Parcelable.Creator<SavedState>() {
            public SavedState createFromParcel(Parcel in) {
                return new SavedState(in);
            }

            public SavedState[] newArray(int size) {
                return new SavedState[size];
            }
        };

    }
}

File: ui/src/main/java/com/seo4d696b75/android/ekisagasu/ui/top/AnimationView.kt
package com.seo4d696b75.android.ekisagasu.ui.top

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.Rect
import android.os.Parcel
import android.os.Parcelable
import android.os.SystemClock
import android.util.AttributeSet
import android.view.View
import android.view.ViewTreeObserver
import com.seo4d696b75.android.ekisagasu.ui.R
import kotlin.math.ceil

/**
 * @author Seo-4d696b75
 * @version 2021/01/10.
 */
class AnimationView :
    View,
    ViewTreeObserver.OnGlobalLayoutListener {
    constructor(context: Context) : this(context, null)

    constructor(context: Context, set: AttributeSet?) : this(context, set, 0)

    constructor(context: Context, set: AttributeSet?, defaultAttr: Int) : super(
        context,
        set,
        defaultAttr,
    ) {

        viewTreeObserver.addOnGlobalLayoutListener(this)

        val res = context.resources
        foreground = BitmapFactory.decodeResource(res, R.drawable.launch_icon)
        background = BitmapFactory.decodeResource(res, R.drawable.loop_line)
        paint = Paint()
        _matrix = Matrix()

        setWillNotDraw(false)
    }

    private var _width: Int = 0
    private var _height: Int = 0
    private var _time: Long = 0
    private var _degree: Float = 0f

    private var running: Boolean = false
    private var requestRun: Boolean = false
    private var requestStop: Boolean = false

    private val background: Bitmap
    private val foreground: Bitmap
    private val _matrix: Matrix
    private val paint: Paint
    private var src: Rect? = null
    private var dst: Rect? = null

    override fun onGlobalLayout() {
        _width = super.getWidth()
        _height = super.getHeight()
        src = Rect(0, 0, foreground.width, foreground.height)
        dst =
            Rect(
                (_width * 0.1f).toInt(),
                (_height * 0.13f).toInt(),
                (_width * 0.8).toInt(),
                (_height * 0.83f).toInt(),
            )
        if (requestRun) runAnimation(true)
    }

    fun runAnimation(run: Boolean) {
        if (running != run) {
            if (running) {
                requestStop = true
            } else {
                if (_width == 0 || _height == 0) {
                    requestRun = true
                    return
                }
                running = true
                requestRun = false
                invalidate()
            }
        }
    }

    override fun onDraw(canvas: Canvas) {
        val dst = this.dst
        val src = this.src
        if (dst == null || src == null) return
        val time = SystemClock.uptimeMillis()
        if (running && _time > 0) {
            var degree = _degree + (time - _time) * 0.01f
            _time = time
            if (requestStop) {
                val step = 360f / 16f
                val th = ceil(_degree / step) * step
                if (degree > th) {
                    running = false
                    requestStop = false
                    _degree = 0f
                    _time = 0L
                }
            }
            if (degree > 360f) {
                degree -= 360f
            }
            _degree = degree
        } else {
            _time = time
        }
        _matrix.reset()
        _matrix.postRotate(_degree, background.width / 2f, background.height / 2f)
        _matrix.postScale(
            _width.toFloat() / background.width,
            _height.toFloat() / background.height,
        )
        canvas.drawBitmap(background, _matrix, paint)
        canvas.drawBitmap(foreground, src, dst, paint)

        if (running) invalidate()
    }

    override fun onRestoreInstanceState(source: Parcelable) {
        val state = source as SavedState
        super.onRestoreInstanceState(state.superState)
        runAnimation(state.mRunning)
    }

    override fun onSaveInstanceState(): Parcelable {
        val superState = super.onSaveInstanceState()
        val state = SavedState(superState)
        state.mRunning = this.running
        return state
    }

    private class SavedState : BaseSavedState {
        var mRunning = false

        constructor(state: Parcelable?) : super(state)
        constructor(source: Parcel) : super(source) {
            val array = BooleanArray(1)
            source.readBooleanArray(array)
            mRunning = array[0]
        }

        override fun writeToParcel(
            out: Parcel,
            flags: Int,
        ) {
            super.writeToParcel(out, flags)
            val array = booleanArrayOf(mRunning)
            out.writeBooleanArray(array)
        }

        companion object {
            @JvmField
            val CREATOR: Parcelable.Creator<SavedState?> =
                object : Parcelable.Creator<SavedState?> {
                    override fun createFromParcel(`in`: Parcel): SavedState = SavedState(`in`)

                    override fun newArray(size: Int): Array<SavedState?> = arrayOfNulls(size)
                }
        }
    }
}
